## Problem Statement

### Business Context

The healthcare industry is rapidly evolving, with professionals facing increasing challenges in managing vast volumes of medical data while delivering accurate and timely diagnoses. The need for quick access to comprehensive, reliable, and up-to-date medical knowledge is critical for improving patient outcomes and ensuring informed decision-making in a fast-paced environment.

Healthcare professionals often encounter information overload, struggling to sift through extensive research and data to create accurate diagnoses and treatment plans. This challenge is amplified by the need for efficiency, particularly in emergencies, where time-sensitive decisions are vital. Furthermore, access to trusted, current medical information from renowned manuals and research papers is essential for maintaining high standards of care.

To address these challenges, healthcare centers can focus on integrating systems that streamline access to medical knowledge, provide tools to support quick decision-making, and enhance efficiency. Leveraging centralized knowledge platforms and ensuring healthcare providers have continuous access to reliable resources can significantly improve patient care and operational effectiveness.

**Questions to Answer**

1. What is the protocol for managing sepsis in a critical care unit?
2. What are the common symptoms of appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?
3. What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?
4. What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?
5. What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

### Objective

As an AI specialist, your task is to develop a RAG-based AI solution using renowned medical manuals to address healthcare challenges. The objective is to **understand** issues like information overload, **apply** AI techniques to streamline decision-making, **analyze** its impact on diagnostics and patient outcomes, **evaluate** its potential to standardize care practices, and **create** a functional prototype demonstrating its feasibility and effectiveness.

### Data Description

The **Merck Manuals** are medical references published by the American pharmaceutical company Merck & Co., that cover a wide range of medical topics, including disorders, tests, diagnoses, and drugs. The manuals have been published since 1899, when Merck & Co. was still a subsidiary of the German company Merck.

The manual is provided as a PDF with over 4,000 pages divided into 23 sections.

## Installing and Importing Necessary Libraries and Dependencies

In [ ]:
# Installation for GPU llama-cpp-python
# uncomment and run the following code in case GPU is being used
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python==0.1.85 --force-reinstall --no-cache-dir -q

# Installation for CPU llama-cpp-python
# uncomment and run the following code in case GPU is not being used
# !CMAKE_ARGS="-DLLAMA_CUBLAS=off" FORCE_CMAKE=1 pip install llama-cpp-python==0.1.85 --force-reinstall --no-cache-dir -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 33.1 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 134.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 266.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.9/16.9 MB 182.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.9/43.9 kB 292.0 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
opencv-contrib-python 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 2.3.2 which is incompatible.
opencv-python-headless 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 2.3.2 which is incompatible.
torch 2.6.0+cu124 requires 

In [ ]:
!pip uninstall -y numpy pandas

!pip install numpy==1.23.5 pandas==1.5.3 -q

Found existing installation: numpy 1.25.2
Uninstalling numpy-1.25.2:
  Successfully uninstalled numpy-1.25.2
Found existing installation: pandas 1.5.3
Uninstalling pandas-1.5.3:
  Successfully uninstalled pandas-1.5.3
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.1/17.1 MB 78.4 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 1.5.3 which is incompatible.
mizani 0.13.5 requires pandas>=2.2.0, but you have pandas 1.5.3 which is incompatible.
opencv-contrib-python 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 1.23.5 which is incompatible.
jax 0.5.2 requires numpy>=1.25, but you have numpy 1.23.5 which is incompatible.
pymc 5.25.0 requires numpy>=1.25.0, but you have numpy 1.23.5 which is incompatible.
albucore 0.0.24 requires numpy>=1.24.4, but you hav

In [ ]:
# For installing the libraries & downloading models from HF Hub
!pip install huggingface_hub==0.23.2 tiktoken==0.6.0 pymupdf==1.25.1 langchain==0.1.1 langchain-community==0.0.13 chromadb==0.4.22 sentence-transformers==2.3.1 -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 3.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 111.9 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 1.5.3 which is incompatible.
mizani 0.13.5 requires pandas>=2.2.0, but you have pandas 1.5.3 which is incompatible.
opencv-contrib-python 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 1.26.4 which is incompatible.
opencv-python-headless 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 1.26.4 which is incompatible.
dask-cudf-cu12 25.6.0 requires pandas<2.2.4dev0,>=2.0, but you have pandas 1.5.3 which is incompatible.
xarray 2025.7.1 requires packaging>=24.1, but you have packaging 23.2 which is incompatible.
xarray 2025.7.1 requires

In [ ]:
#Libraries for processing dataframes,text
import json,os
import tiktoken
import pandas as pd

#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

#Libraries for downloading and loading the llm
from huggingface_hub import hf_hub_download
from llama_cpp import Llama

## Question Answering using LLM

#### Downloading and Loading the model

In [ ]:
# Defining the Hugging Face repository and model version for Mistral-7B fine-tuned for instruction-following
model_name_or_path = 'TheBloke/Mistral-7B-Instruct-v0.2-GGUF'

# Specifying the file name for the quantized Mistral-7B model in GGUF format (Q6_K for optimal performance)
model_basename = 'mistral-7b-instruct-v0.2.Q6_K.gguf'

In [ ]:
# Downloading the specified model file from Hugging Face Hub and store its local path
model_path = hf_hub_download(
    repo_id=model_name_or_path, #The Hugging Face repository containing the model
    filename=model_basename  # The specific model file to download (in GGUF format)
)
#The GGUF format is used because it provides memory-efficient storage and faster inference while maintaining compatibility across different hardware platforms.

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


mistral-7b-instruct-v0.2.Q6_K.gguf:   0%|          | 0.00/5.94G [00:00<?, ?B/s]

In [ ]:
# Loading the LLaMA model with specified context, GPU layers, and batch size
llm = Llama(
    model_path=model_path, #Path to the GGUF model file
    n_ctx=2300, #Sets the context window to 2300 tokens (how much text the model can "see" at once)
    n_gpu_layers=38, #Loads 38 model layers onto GPU for faster inference (set to 0 for CPU-only)
    n_batch=512 #Number of tokens processed at once
)

AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | 


#### Response

In [ ]:
def response(query,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    model_output = llm(
      prompt=query,
      max_tokens=max_tokens,
      temperature=temperature,
      top_p=top_p,
      top_k=top_k
    )

    return model_output['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
response('What is the protocol for managing sepsis in a critical care unit?')

'\n\nSepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical care unit. The following are general steps for managing sepsis in a critical care unit:\n\n1. Early recognition: Recognize the signs and symptoms of sepsis early and initiate treatment as soon as possible. Sepsis can present with various clinical features, including fever or hypothermia, tachycardia or bradycardia, altered mental status, respiratory distress, and lactic acidosis.\n2. Resusc'

#### Observations
* Starts with a medically accurate definition of sepsis, which sets the context well.

* Emphasizes early recognition and intervention, which aligns with standard critical care protocols.

* Appears to follow a logical structure, beginning with recognition and potentially continuing to treatment (though the response is truncated).


### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
response('What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?')

Llama.generate: prefix-match hit


'\n\nAppendicitis is a medical condition characterized by inflammation of the appendix, a small tube-shaped organ located in the lower right side of the abdomen. The symptoms of appendicitis can vary from person to person, but some common signs include:\n\n1. Abdominal pain: The pain may start as a mild discomfort around the navel or in the lower right abdomen, which then gradually moves to the right lower quadrant and becomes more severe over time. The pain may be constant or intermittent and is often worsened by movement, coughing, or deep breathing'

#### Observations

* Provides a clear and accurate definition of appendicitis, including the anatomical location of the appendix.

* Covers only the symptom aspect of the query; does not address the part asking whether it can be treated with medicine or what surgical procedure is used.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
response('What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?')

Llama.generate: prefix-match hit


"\n\nSudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles. It can result in round or oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, or eyelashes.\n\nThe exact cause of alopecia areata is not known, but it's believed to be related to a problem with the immune system. Some possible triggers for this condition include stress, genetics, viral infections, and certain medications."

#### Observations

* The response correctly identifies alopecia areata as the condition behind sudden patchy hair loss and It mentions several possible causes (stress, genetics, viral infections, medications), providing good contextual background on the condition.

* However, it fails to address the treatment part of the query, leaving the user's question only partially answered.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
response('What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?')

Llama.generate: prefix-match hit


"\n\nA person who has sustained a physical injury to brain tissue, also known as a traumatic brain injury (TBI), may require various treatments depending on the severity and location of the injury. Here are some common treatments recommended for TBIs:\n\n1. Emergency care: The first priority is to ensure the person's airway is clear, they are breathing, and their heart is beating normally. In severe cases, emergency surgery may be required to remove hematomas or other obstructions.\n2. Medications: Depending on the symptoms, medications may be prescribed to manage conditions such as"

### Observations

* The response identifies the physical injury to brain as TBI medically.
* It also provides couple of common treatments for the same.
* Also it provides the context that what kind of medications should be provided based on symptoms.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
response('What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?')

Llama.generate: prefix-match hit


"\n\nFirst and foremost, if you suspect that someone has fractured their leg while hiking, it's essential to ensure their safety and prevent further injury. Here are some necessary precautions:\n\n1. Keep the person calm and still: Encourage them to remain as still as possible to minimize pain and prevent worsening the injury.\n2. Assess the situation: Check for any signs of shock, such as pale skin, rapid heartbeat, or shallow breathing. If you notice these symptoms, seek medical help immediately.\n3. Immobilize the leg: Use a splint, sl"

#### Observations

* The response provides the necessary precautions.
* The next step logically talks about the treatment.
* However the steps on their care and recovery is not answered.

### Pros and Cons of the Mistral 7B Model as a Medical Assistant
* Pros
  * Comprehensive and Medically Sound Responses
  * The Mistral 7B model demonstrates strong capabilities in delivering detailed and clinically appropriate answers, effectively addressing a wide range of medical queries.
  * Responses are well-organized, often presented in numbered or bullet-point formats, making them easy to follow and digest for both medical professionals and general users.
  * The model accurately interprets the intent of the input and generates relevant, targeted information that aligns closely with the user's question.

* Cons
  * Truncated or Incomplete Responses
  * Many answers are cut off mid-sentence (e.g., responses regarding treatments for traumatic brain injuries or fractures), likely due to token limitations. This can result in partial or fragmented information, reducing the overall usefulness of the answer.
  * Overly Generic Output
  * While structurally sound, some responses lack depth and specificity. This generic nature may reduce the effectiveness of the model in scenarios requiring detailed, patient-specific recommendations.
  * The model often provides high-level advice or best practices rather than nuanced, case-specific insights. This limits its applicability in more complex or variable medical scenarios.

## Question Answering using LLM with Prompt Engineering

In [ ]:
def generate_llm_response_with_prompt1(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=1024,
        temperature=0.7,
        top_p=0.95,
        repeat_penalty=1.1,
        top_k=50,
        stop=['</s>'],
        echo=False,
    )

    # Return only the generated answer text
    return response["choices"][0]["text"].strip()

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
generate_llm_response_with_prompt1("What is the protocol for managing sepsis in a critical care unit?")

Llama.generate: prefix-match hit


'1. Early recognition: Identify sepsis suspects within 1 hour of suspected onset and initiate treatment as soon as possible.\n2. Fluid resuscitation: Aim for adequate tissue perfusion, typically using crystalloids to restore intravascular volume.\n3. Vasopressors: For persistent hypotension despite fluid resuscitation, administer vasopressors to maintain mean arterial pressure (MAP) ≥65 mmHg.\n4. Antibiotics: Administer broad-spectrum antibiotics based on culture results or suspected organism(s).\n5. Source control: Identify and address the source of infection as soon as possible, which may include surgery, drainage, or debridement.\n6. Corticosteroids: Consider using corticosteroids in patients with septic shock who have not responded to fluid resuscitation and vasopressors within 1 hour.\n7. Respiratory support: Provide appropriate ventilatory support as needed, including non-invasive or invasive mechanical ventilation.\n8. Inotropes: Administer inotropic agents to improve cardiac ou

Modified Parameter Temperature to 0.0

In [ ]:
def generate_llm_response_with_prompt2(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=1024,
        temperature=0.0,
        top_p=0.95,
        repeat_penalty=1.1,
        top_k=50,
        stop=['</s>'],
        echo=False,
    )

    # Return only the generated answer text
    return response["choices"][0]["text"].strip()

In [ ]:
generate_llm_response_with_prompt2("What is the protocol for managing sepsis in a critical care unit?")

Llama.generate: prefix-match hit


"The management of sepsis in a critical care unit involves early recognition, prompt initiation of antibiotics, fluid resuscitation, and supportive measures. The Sequential [Sepsis-related] Organ Failure Assessment (SOFA) score can be used to identify patients at risk for developing sepsis or worsening sepsis. Once sepsis is suspected, blood cultures should be obtained and broad-spectrum antibiotics administered based on the patient's clinical presentation and local microbiology data. Fluid resuscitation with crystalloids or colloids may be necessary to maintain adequate tissue perfusion. Vasopressors and/or mechanical ventilation may also be required for hemodynamic support. Corticosteroids, anticoagulation, and other adjunctive therapies may be considered based on individual patient needs and clinical trials data. Close monitoring of vital signs, laboratory values, and organ function is essential to guide therapy and prevent progression to septic shock or multiple organ failure."

#### Observations

* Response with (Temperature = 0.0)

  * Highly Deterministic Output: The model provides a focused, reliable, and fact-based response—ideal for clinical applications where consistency is critical.

  * Paragraph Style: Presents a comprehensive answer in a flowing narrative rather than a bullet format, suitable for explanatory contexts.

  * Detailed but Compact: Includes relevant concepts like SOFA score, microbiology data, and adjunctive therapies, showing depth without verbosity.

* Response with (Temperature = 0.7)

  * More Creative/Explorative Output: The model uses a structured, list-based format with detailed treatment steps, possibly influenced by higher randomness.

  * Clear Actionable Format: Breaks down the protocol step-by-step (e.g., vasopressors, antibiotics, corticosteroids), making it easy to follow in a procedural setting.

  * Risk of Inconsistency: Slight variations in phrasing and sequencing may occur across runs due to higher temperature, which could affect repeatability.


Overall Response with (Temperature = 0.7) is better and superior.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
generate_llm_response_with_prompt1("What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?")

Llama.generate: prefix-match hit


'The symptoms of appendicitis include:\n1. Sudden pain in the lower right abdomen that may start as a dull ache but progresses to sharp cramping and localized tenderness.\n2. Loss of appetite, nausea, and vomiting.\n3. Fever, often low-grade at first but rising as the condition worsens.\n4. Abdominal swelling and rigidity.\n5. Constipation or diarrhea.\n\nAppendicitis cannot be cured with medication alone since the appendix is a hollow, tube-like organ that can become inflamed, filled with pus, and eventually burst if left untreated. If the appendix ruptures, peritonitis - an infection of the abdominal cavity - may occur, leading to severe complications like sepsis.\n\nTo prevent rupture and treat appendicitis, surgical intervention is necessary. The standard procedure for treating appendicitis is an appendectomy, which involves removing the inflamed appendix through a small incision in the abdomen or through laparoscopic surgery with smaller incisions. Surgery should be performed as s

Modified top_p parameter to 0.85

In [ ]:
def generate_llama_response_with_prompt3(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=1024,
        temperature=0.7,
        top_p=0.85,
        repeat_penalty=1.1,
        top_k=50,
        stop=['</s>'],
        echo=False,
    )

    # Return only the generated answer text
    return response["choices"][0]["text"].strip()

In [ ]:
generate_llama_response_with_prompt3("What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?")

Llama.generate: prefix-match hit


"The symptoms of appendicitis typically include:\n1. Sudden pain in the lower right abdomen that may start near the navel and then shift to the right side.\n2. Loss of appetite and/or feeling sick to your stomach.\n3. Fever, which may be low-grade at first but can rise as high as 101 degrees Fahrenheit (38.3 degrees Celsius).\n4. Vomiting and diarrhea or constipation.\n5. Abdominal swelling and rigidity, particularly in the right lower quadrant of the abdomen.\n6. Pain when moving your body, coughing, or taking deep breaths.\n7. Inability to pass gas or have a bowel movement.\n\nAppendicitis cannot be cured with medicine alone as the appendix may burst and release infectious materials into the abdominal cavity if left untreated. The standard treatment for appendicitis is surgical removal of the appendix, called an appendectomy. This procedure can typically be performed laparoscopically, which involves making small incisions in the abdomen and using a camera and specialized instruments 

#### Observations

* Response with top_p=0.95

  * More Generalized Language: The explanation is medically sound but avoids specific numeric values or elaboration (e.g., just says "fever, often low-grade" without temperature).

  * Compact & Concise: Shorter, less elaborated steps and symptoms—more streamlined for a general overview.

  * More Focused on Key Symptoms: Prioritizes core symptoms and treatment steps without digressing into edge-case details.

  * Slightly Less Detail-Oriented: Skips certain specific clinical scenarios (e.g., doesn’t mention gas or bowel movement issues).

* Response with top_p=0.85

  * More Detailed & Specific: Includes exact fever measurements (e.g., "as high as 101°F / 38.3°C") and less common symptoms like inability to pass gas.

  * Richer Clinical Description: Elaborates on how pain worsens with movement or deep breathing—demonstrates broader sampling from the model's probability space.

  * Increased Verbosity: Slightly longer and more exhaustive, potentially useful in clinical settings where detailed differentiation matters.

  * Enhanced Realism: Refers to technical terms (e.g., "laparoscopically," "specialized instruments") for surgical explanation, offering a more grounded medical portrayal.


Overall Response with top_p=0.85 is better and superior, Since it captures broader and more nuanced symptomology and diagnosis-related cues.


### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
generate_llm_response_with_prompt1("What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?")

Llama.generate: prefix-match hit


"Sudden patchy hair loss can be distressing, but there are several potential causes and treatments. Some common causes include:\n\n1. Alopecia Areata: This is an autoimmune condition where your immune system mistakenly attacks hair follicles, leading to hair loss in small patches on the scalp, face, or other areas of the body. There is no definitive cure for alopecia areata, but treatments like corticosteroids (topical or injectable), minoxidil, and anthralin can help stimulate hair regrowth.\n2. Traction Alopecia: This occurs when excessive pulling or tension is applied to the hair follicles due to tight hairstyles (like braids, ponytails, or weaves) or constant rubbing against a fabric or other surface. The best way to treat traction alopecia is by avoiding the causing factor and giving your scalp time to heal.\n3. Nutritional Deficiencies: A lack of essential nutrients can lead to hair loss. Ensure you have a balanced diet rich in protein, iron, zinc, vitamin D, and biotin. If neede

Modified parameter is top_k=80

In [ ]:
def generate_llama_response_with_prompt4(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=1024,
        temperature=0.7,
        top_p=0.95,
        repeat_penalty=1.1,
        top_k=80,
        stop=['</s>'],
        echo=False,
    )

    # Return only the generated answer text
    return response["choices"][0]["text"].strip()

In [ ]:
generate_llama_response_with_prompt4("What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?")

Llama.generate: prefix-match hit


"Sudden patchy hair loss can have various causes, including the following:\n\n1. Alopecia Areata: An autoimmune disease that results in hair loss, typically presenting as round or oval patches of baldness on the scalp. Treatment options include topical corticosteroids, immunotherapy (such as minoxidil), or intralesional triamcinolone injections.\n2. Trauma or stress: Physical or emotional stress can cause temporary hair loss, known as telogen effluvium. This condition usually resolves on its own within six months. Treatment may include managing the underlying cause and encouraging a healthy lifestyle through proper diet, exercise, and stress reduction techniques.\n3. Nutritional deficiencies: Lack of essential nutrients such as iron, zinc, or biotin can lead to hair loss. A balanced diet rich in these nutrients should be considered. In some cases, supplements may be necessary under the guidance of a healthcare professional.\n4. Hormonal imbalances: Hormonal fluctuations, particularly t

#### Observations

* Response with top_k = 50
  * More Direct and Concise: Focuses quickly on key causes like alopecia areata, traction alopecia, and nutritional deficiency.

  * Simpler Terminology: Easier for general users to read and understand; avoids overly technical phrasing.

  * Practical Advice: Emphasizes actionable treatment options like minoxidil, corticosteroids, and dietary recommendations.

  * Slightly Less Depth: Though informative, it provides fewer distinct causes or classifications compared to the higher top_k.

* Response with top_k = 80
  * Broader Coverage: Mentions more causes (e.g., telogen effluvium, hormonal imbalances, trauma/stress) with slightly more medical detail.

  * More Formal Style: Uses more clinical or technical phrasing, suitable for a medically informed audience.

  * Higher Informational Richness: Includes treatment categories (e.g., immunotherapy, intralesional injections, supplements) and duration (e.g., “resolves within six months”).

  * Slightly Wordier: Could be harder for laypeople to follow without summarization.

Overall the response with top_k=80 seems to be better as it has greater variety and depth in causes/treatments.  

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
generate_llm_response_with_prompt1("What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?")

Llama.generate: prefix-match hit


'The treatment for a brain injury depends on the severity and location of the injury. For mild injuries such as concussions, rest, hydration, and avoiding physical or cognitive activities that may worsen symptoms are usually recommended. More severe injuries may require hospitalization, surgery, medication to manage symptoms like swelling or seizures, rehabilitation therapy, and assistive devices. Rehabilitation therapies may include physical therapy, occupational therapy, speech-language therapy, and cognitive rehabilitation. In some cases, ongoing care and support from healthcare professionals and family members may be necessary. It is important to note that recovery from a brain injury can be a long process, and individual progress may vary greatly.'

Modified Parameters is max_tokens = 512

In [ ]:
def generate_llama_response_with_prompt5(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=512,
        temperature=0.7,
        top_p=0.95,
        repeat_penalty=1.1,
        top_k=50,
        stop=['</s>'],
        echo=False,
    )

    # Return only the generated answer text
    return response["choices"][0]["text"].strip()

In [ ]:
generate_llama_response_with_prompt5("What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?")

Llama.generate: prefix-match hit


'The treatment for a brain injury depends on its severity and the specific areas of the brain that have been affected. Generally, the following interventions may be recommended:\n\n1. Emergency care: For severe brain injuries, immediate medical attention is required to stabilize the person and prevent further damage. This might include surgery, intubation, and administration of medications to manage swelling or seizures.\n2. Rehabilitation: Rehabilitation programs can help individuals regain lost skills and improve their overall function. Depending on the location and extent of the injury, a rehabilitation team may include physical therapists, occupational therapists, speech-language pathologists, and neuropsychologists.\n3. Medications: Various medications might be prescribed to manage symptoms associated with brain injuries, such as pain, seizures, or depression.\n4. Assistive devices: Devices like wheelchairs, walkers, or communication aids may be necessary to help individuals with 

#### Observations

* Response with max_tokens = 1024
  * More Concise and Focused: Despite the higher token limit, the response remains compact and avoids unnecessary elaboration.

  * Highlights Key Therapies: Mentions important treatment types—cognitive rehab, speech therapy, physical and occupational therapy.

  * Organized but Less Structured: Delivered in paragraph form, which may reduce readability for complex information.

  * General Guidance: Offers balanced info for both mild and severe injuries but lacks numbered breakdown or detailed categorization.

* Response with max_tokens = 512
  * More Structured Output: Uses numbered format (1–4) to present different intervention types like emergency care, rehab, medication, and assistive devices.

  * Better Clarity and Readability: Easier to scan and digest; ideal for readers looking for quick takeaways.

  * Deeper Specificity: Gives more clinical detail—mentions intubation, neuropsychologists, and condition-specific devices.

  * Compact Yet Informative: Despite the lower token limit, it packs in more layers of medical and practical depth.

Overall Response with max_tokens=1024 is better and superior as with more tokens we can get more informed about the situation.  

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
generate_llm_response_with_prompt1("What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?")

Llama.generate: prefix-match hit


"1. Immediate Care: If a hiker suspects they have a fractured leg, they should not move the limb to avoid causing further damage. Instead, they should try to keep the leg stable by immobilizing it with a makeshift splint or using clothing to create a sling if possible.\n2. Seek Medical Assistance: It's crucial to seek medical attention as soon as possible, especially if there are signs of severe pain, swelling, deformity, or numbness in the leg. In remote areas without immediate access to emergency services, it may be necessary to hike out with the assistance of others or call for a rescue team.\n3. Treatment: A fractured leg will typically require immobilization using a cast or brace for several weeks. The healthcare provider may also prescribe pain medication and physical therapy exercises to help restore mobility and strength to the affected limb.\n4. Recovery: Proper recovery from a fractured leg involves following the healthcare professional's instructions for care, attending all 

Modified Parameters are :-
1. temperature = 0.7

2. top_p = 0.85

3. top_k = 80

4. max_tokens = 1024

In [ ]:
def generate_llama_response_with_prompt6(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=1024,
        temperature=0.7,
        top_p=0.85,
        repeat_penalty=1.1,
        top_k=80,
        stop=['</s>'],
        echo=False,
    )

    # Return only the generated answer text
    return response["choices"][0]["text"].strip()

In [ ]:
generate_llama_response_with_prompt6("What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?")

Llama.generate: prefix-match hit


"First and foremost, if you suspect that someone has fractured their leg while hiking, it's crucial to ensure their safety and prevent further injury. Here are the necessary steps:\n\n1. Keep the person calm and still to reduce pain and swelling.\n2. Assess the severity of the injury by checking for signs of open wounds, deformities, or inability to move the leg.\n3. If the fracture is suspected to be severe or open, do not attempt to move the person without proper medical assistance. Call emergency services and wait for medical help to arrive.\n4. For less severe fractures, provide a makeshift splint using materials like sticks, clothing, or hiking poles to immobilize the leg to prevent further movement.\n5. Once the leg is stabilized, try to keep the person warm by covering them with blankets or additional clothing.\n6. Offer water or other fluids to help keep them hydrated and provide any necessary first aid supplies, such as bandages or pain relief medication if available.\n7. If p

#### Observations

* Response with generate_llama_response_with_prompt6

  * Highly structured and step-wise (7 well-separated points)

  * Includes more realistic field instructions like:

    * Immobilization with poles

    * Avoiding movement without assistance

  * Keeping the injured person warm

  * Offering water and basic supplies

  * Language tone is measured, cautious, and situationally aware

  * Better aligns with realistic emergency context

* Response with generate_llama_response_with_prompt1

  * More medicalized and authoritative tone

  * Structured into 4 phases: Immediate care, medical assistance, treatment, and recovery

  * Better continuity and readability as a short summary  

Overall due to richer, more pragmatic content, and better step-by-step clarity — ideal for real-world emergency aid scenario is Response with generate_llama_response_with_prompt6

## Data Preparation for RAG

### Loading the Data

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
merck_pdf_path = '/content/drive/MyDrive/Colab Notebooks/medical_diagnosis_manual.pdf'
pdf_loader = PyMuPDFLoader(merck_pdf_path)
merck = pdf_loader.load()

### Data Overview

#### Checking the first 5 pages

In [ ]:
for i in range(5):
    print(f"Page Number : {i+1}",end="\n")
    print(merck[i].page_content,end="\n")

Page Number : 1
vishnukv327@gmail.com
5NQRJADWON
ant for personal use by vishnukv327@gm
shing the contents in part or full is liable 

Page Number : 2
vishnukv327@gmail.com
5NQRJADWON
This file is meant for personal use by vishnukv327@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.

Page Number : 3
Table of Contents
1
Front    ................................................................................................................................................................................................................
1
Cover    .......................................................................................................................................................................................................
2
Front Matter    ..........................................................................................................................................................................................

#### Checking the number of pages

In [ ]:
len(merck)

4114

### Data Chunking

In [ ]:
# Initializing a RecursiveCharacterTextSplitter to split the text into manageable chunks for embedding and retrieval
text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    encoding_name='cl100k_base',
    chunk_size=512,
    chunk_overlap= 20
)

In [ ]:
#loading the PDF document, extracting its text, and splitting it into smaller chunks
document_chunks = pdf_loader.load_and_split(text_splitter)

In [ ]:
#Checking the number of text chunks the pdf has been split into
len(document_chunks)

8492

Lets validate that there is overalapping b/w chunks

In [ ]:
document_chunks[120].page_content

'Specific indications for enteral nutrition include the following:\n• Prolonged anorexia\n• Severe protein-energy undernutrition\nThe Merck Manual of Diagnosis & Therapy, 19th Edition\nChapter 3. Nutritional Support\n70\nvishnukv327@gmail.com\n5NQRJADWON\nThis file is meant for personal use by vishnukv327@gmail.com only.\nSharing or publishing the contents in part or full is liable for legal action.'

In [ ]:
document_chunks[119].page_content

'Table 3-1).\nAssessing Response to Nutritional Support\nThere is no gold standard to assess response. Clinicians commonly use indicators of lean body mass\nsuch as the following:\n• Body mass index (BMI)\n• Body composition analysis\n• Body fat distribution (see pp.\n11 and 58)\nNitrogen balance, response to skin antigens, muscle strength measurement, and indirect calorimetry can\nalso be used.\n[Table 3-1. Estimated Adult Daily Protein Requirement]\nNitrogen balance, which reflects the balance between protein needs and supplies, is the difference\nbetween amount of nitrogen ingested and amount lost. A positive balance (ie, more ingested than lost)\nimplies adequate intake. Precise measurement is impractical, but estimates help assess response to\nnutritional support. Nitrogen intake is estimated from protein intake: nitrogen (g) equals protein (g)/6.25.\nEstimated nitrogen losses consist of urinary nitrogen losses (estimated by measuring urea nitrogen\ncontent of an accurately obtain

In [ ]:
document_chunks[121].page_content

"• Coma or depressed sensorium\n• Liver failure\n• Inability to take oral feedings due to head or neck trauma or neurologic disorders\n• Critical illnesses (eg, burns) causing metabolic stress\nOther indications may include bowel preparation for surgery in seriously ill or undernourished patients,\nclosure of enterocutaneous fistulas, and small-bowel adaptation after massive intestinal resection or in\ndisorders that may cause malabsorption (eg, Crohn's disease).\nProcedure: If tube feeding is needed for ≤ 4 to 6 wk, a small-caliber, soft nasogastric or nasoenteric (eg,\nnasoduodenal) tube made of silicone or polyurethane is usually used. If a nasal injury or deformity makes\nnasal placement difficult, an orogastric or other oroenteric tube can be placed.\nTube feeding for > 4 to 6 wk usually requires a gastrostomy or jejunostomy tube, placed endoscopically,\nsurgically, or radiologically. Choice depends on physician capabilities and patient preference.\nJejunostomy tubes are useful fo

### Embedding

In [ ]:
#This model is chosen because of its embedding vector size is the same as our token size in chunking (512).
embedding_model = SentenceTransformerEmbeddings(model_name='thenlper/gte-large')

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

modules.json:   0%|          | 0.00/385 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/619 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/670M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/342 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

In [ ]:
# Generating embedding for the first document chunk
embedding_1 = embedding_model.embed_query(document_chunks[0].page_content)
# Generating embedding for the second document chunk
embedding_2 = embedding_model.embed_query(document_chunks[1].page_content)

In [ ]:
#Checking if both are of the same size
print("Dimension of the embedding vector ",len(embedding_1))
len(embedding_1)==len(embedding_2)

Dimension of the embedding vector  1024


True

### Vector Database

In [ ]:
import os
# Creating the output directory 'merck_db' if it doesn't already exist, so we can save the processed data or vector database files there.
out_dir = 'merck_db'

if not os.path.exists(out_dir):
  os.makedirs(out_dir)

In [ ]:
vectorstore = Chroma.from_documents( #creating a Chroma vector store from a set of document chunks.
    document_chunks, #creating a list of text chunks that will be converted into embeddings..
    embedding_model, #model responsible for embedding the document chunks into vector representations
    persist_directory=out_dir #name of the collection in the Chroma database
)

ERROR:chromadb.telemetry.product.posthog:Failed to send telemetry event ClientStartEvent: capture() takes 1 positional argument but 3 were given
ERROR:chromadb.telemetry.product.posthog:Failed to send telemetry event ClientCreateCollectionEvent: capture() takes 1 positional argument but 3 were given


In [ ]:
 #Loading Chroma vector store with the given embedding model
vectorstore = Chroma(persist_directory=out_dir,embedding_function=embedding_model)

ERROR:chromadb.telemetry.product.posthog:Failed to send telemetry event ClientStartEvent: capture() takes 1 positional argument but 3 were given
ERROR:chromadb.telemetry.product.posthog:Failed to send telemetry event ClientCreateCollectionEvent: capture() takes 1 positional argument but 3 were given


In [ ]:
#Accessing the embedding function used in the Chroma vector store
vectorstore.embeddings

HuggingFaceEmbeddings(client=SentenceTransformer(
  (0): Transformer({'max_seq_length': 512, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 1024, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False})
  (2): Normalize()
), model_name='thenlper/gte-large', cache_folder=None, model_kwargs={}, encode_kwargs={}, multi_process=False)

In [ ]:
#Performing a similarity search in the vector store to find the top 3 most similar documents to "Wheezing"
vectorstore.similarity_search("Wheezing",k=3)

[Document(page_content='obstruction; and avoiding asthma misdiagnosis and treatment.\nVocal cord dysfunction associated with psychiatric diagnoses is often resistant to these measures.\nReferral for psychiatric counseling is indicated in these cases.\nWheezing\nWheezing is a relatively high-pitched whistling noise produced by movement of air through narrowed or\ncompressed small airways. It is a symptom as well as a physical finding.\nPathophysiology\nAirflow through a narrowed or compressed segment of a small airway becomes turbulent, causing\nvibration of airway walls; this vibration produces the sound of wheezing.\nWheezes are more common during expiration because increased intrathoracic pressure during this phase\nnarrows the airways. Wheezing during expiration alone indicates milder obstruction than wheezing during\nboth inspiration and expiration, which suggests more severe airway narrowing.\nBy contrast, turbulent flow of air through a narrowed segment of the large, extrathoraci

In [ ]:
#Performing a similarity search in the vector store to find the top 3 most similar documents to "Alopecia Areata"
vectorstore.similarity_search("Alopecia Areata ",k=3)

[Document(page_content='infection or emotional stress. It occasionally coexists with autoimmune vitiligo or thyroiditis.\nDiagnosis\n• Examination\nDiagnosis is by inspection. Alopecia areata typically manifests as discrete circular patches of hair loss\ncharacterized by short broken hairs at the margins, which resemble exclamation points. Nails are\nsometimes pitted or display trachyonychia, a roughness of the nail also seen in lichen planus. Differential\ndiagnosis includes tinea capitis, trichotillomania, discoid lupus, and secondary syphilis. Measures of\nthyroid-stimulating hormone, vitamin B12, and autoantibodies are indicated only when coexisting disease\nis suspected.\nTreatment\n• Corticosteroids\n• Sometimes topical anthralin, minoxidil or both\nTreatment is with corticosteroids. Triamcinolone acetonide suspension (in doses not to exceed 0.1 mL per\ninjection site, eg, 10 mg/mL concentration to deliver 1 mg) can be injected intradermally if the lesions are\nsmall. Potent topi

#### Observations

In both cases where we have retrieved the chunks we can observe that all the chunks are related to the terms wheezing and Alopecia Areata.

### Retriever

In [ ]:
retriever = vectorstore.as_retriever( #Converting the Chroma vector store into a retriever for querying.
    search_type='similarity', #Specifying that retrieval is based on cosine similarity
    search_kwargs={'k': 3} #Retrieving the top 3 most similar documents for a given query.
)


In [ ]:
user_input = 'What are the symptoms of migraine?'
rel_docs = retriever.get_relevant_documents(user_input)
rel_docs

[Document(page_content='are less common than visual auras. Some patients have an aura with little or no headache.\nHeadache varies from moderate to severe, and attacks last from 4 hours to several days, typically\nresolving with sleep. The pain is often unilateral but may be bilateral, most often in a frontotemporal\ndistribution, and is typically described as pulsating or throbbing.\nMigraine is more than a headache. Associated symptoms such as nausea (and occasionally vomiting),\nphotophobia, sonophobia, and osmophobia are prominent. Patients report difficulty concentrating during\nattacks. Routine physical activity usually aggravates migraine headache; this effect, plus the photophobia\nand sonophobia, encourages most patients to lie in a dark, quiet room during attacks. Severe attacks can\nThe Merck Manual of Diagnosis & Therapy, 19th Edition\nChapter 178. Headache\n1886\nvishnukv327@gmail.com\n5NQRJADWON\nThis file is meant for personal use by vishnukv327@gmail.com only.\nSharing 

### System and User Prompt Template

In [ ]:
# System message instructing the LLM to only answer using Merck Manual 19th Edition
qna_system_message = """
You are a helpful assistant trained to answer questions based only on the Merck Manual of Medical Diagnosis and Therapy, Nineteenth Edition.
Use the context provided to find accurate and reliable answers.
If the answer is not found in the context, reply with "I don't know".
Do not mention the context or the Merck Manual in your final answer.
"""

In [ ]:
# Template for formatting the user's input with context from the Merck Manual, 19th Edition and the actual medical question.
qna_user_message_template = """
###Context
The following excerpts are from the Merck Manual of Medical Diagnosis and Therapy, Nineteenth Edition:
{context}

###Question
{question}
"""

### Response Function

In [ ]:
def generate_rag_response(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

## Question Answering using RAG

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
user_input1 = 'What is the protocol for managing sepsis in a critical care unit?'
print(generate_rag_response(user_input1))

Llama.generate: prefix-match hit


###Answer
The protocol for managing sepsis in a critical care unit includes prompt empiric antibiotic therapy, source control such as drainage of abscesses or surgical excision of necrotic tissues, normalization of blood glucose levels with continuous IV insulin infusion, corticosteroid therapy with replacement doses, and consideration of emerging therapies like activated protein C for severe sepsis. Antibiotics should be selected based on the suspected source, clinical setting, and sensitivity patterns common to that specific inpatient unit. Vancomycin must be added if


#### Observations

The Response in this case is succinct and answers the question with correct info.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
generate_rag_response("What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?")

Llama.generate: prefix-match hit


"The common symptoms for appendicitis include abdominal pain, anorexia (loss of appetite), and abdominal tenderness. Appendicitis cannot be cured via medicine alone as it requires surgical removal of the appendix. The treatment is called appendectomy, which can be performed either open or laparoscopically. Antibiotics are also administered before the surgery to prevent infection. If the appendix is perforated, antibiotics should be continued until the patient's temperature and white blood cell count have normalized. In cases where a large inflammatory mass is found"

#### Observations

The Response in this case covers all the required details also mentioned above the surgical process in detail.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
generate_rag_response("What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?")

Llama.generate: prefix-match hit


'Based on the context provided, the effective treatments or solutions for addressing sudden patchy hair loss include:\n1. Corticosteroids (topical, intralesional, or systemic)\n2. Minoxidil (topical)\n3. Anthralin (topical)\n4. Immunotherapy (diphencyprone or squaric acid dibutylester)\n5. Psoralen plus ultraviolet A (PUVA)\n6. Hormonal modulators such as oral contraceptives or spironolactone for female-pattern hair'

#### Observations

The response for this query in detail provides the effective treatment and solutions but it doesn't provide info on what could be the possible causes behind it. The possible reason of missing that could be because of not having enough max_tokens for the response to be more descriptive.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
generate_rag_response("What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?")

Llama.generate: prefix-match hit


'###Answer\nThe initial treatment includes ensuring a reliable airway and maintaining adequate ventilation, oxygenation, and blood pressure. Surgery may be needed in patients with more severe injuries to place monitors to track and treat intracranial pressure, decompress the brain if intracranial pressure is increased, or remove intracranial hematomas. In the first few days after the injury, maintaining adequate brain perfusion and oxygenation and preventing complications of altered sensorium are important. Subsequently, many patients require rehabilitation. There is no specific treatment for traumatic brain injury, but supportive care'

#### Observations

The response for this query covers the complete steps from intial treatment to  post treatment care.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
generate_rag_response("What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?")

Llama.generate: prefix-match hit


'Based on the context provided, the following are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip:\n1. Elevate the injured limb above heart level for the first 2 days to minimize swelling and allow gravity to help drain edema fluid.\n2. Apply elastic bandages (batting, and layers 2 and 4) to the injured area after the initial 2 days.\n3. Periodically apply warmth using a heating pad for 15 to 20 minutes to relieve pain and speed healing.\n4.'

#### Observations

The response for this query covers the required precautions, treatment steps and also the recovery.

### Fine-tuning

### Query 1: What is the protocol for managing sepsis in a critical care unit?

Modified chunk_overlap to 0, So that there is no chunk overlapping present.

In [ ]:
# Initializing a RecursiveCharacterTextSplitter to split the text into manageable chunks for embedding and retrieval
text_splitter1 = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    encoding_name='cl100k_base',
    chunk_size=512,
    chunk_overlap= 0
)

In [ ]:
#loading the PDF document, extracting its text, and splitting it into smaller chunks
document_chunks1 = pdf_loader.load_and_split(text_splitter1)

In [ ]:
#Checking the number of text chunks the pdf has been split into
len(document_chunks1)

8461

In [ ]:
# Generating embedding for the first document chunk
embedding_1 = embedding_model.embed_query(document_chunks1[0].page_content)
# Generating embedding for the second document chunk
embedding_2 = embedding_model.embed_query(document_chunks1[1].page_content)


In [ ]:
vectorstore1 = Chroma.from_documents( #creating a Chroma vector store from a set of document chunks.
    document_chunks1, #creating a list of text chunks that will be converted into embeddings..
    embedding_model, #model responsible for embedding the document chunks into vector representations
    persist_directory=out_dir #name of the collection in the Chroma database
)

ERROR:chromadb.telemetry.product.posthog:Failed to send telemetry event ClientStartEvent: capture() takes 1 positional argument but 3 were given
ERROR:chromadb.telemetry.product.posthog:Failed to send telemetry event ClientCreateCollectionEvent: capture() takes 1 positional argument but 3 were given


In [ ]:
 #Loading Chroma vector store with the given embedding model
vectorstore1 = Chroma(persist_directory=out_dir,embedding_function=embedding_model)

ERROR:chromadb.telemetry.product.posthog:Failed to send telemetry event ClientStartEvent: capture() takes 1 positional argument but 3 were given
ERROR:chromadb.telemetry.product.posthog:Failed to send telemetry event ClientCreateCollectionEvent: capture() takes 1 positional argument but 3 were given


In [ ]:
#Accessing the embedding function used in the Chroma vector store
vectorstore1.embeddings

HuggingFaceEmbeddings(client=SentenceTransformer(
  (0): Transformer({'max_seq_length': 512, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 1024, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False})
  (2): Normalize()
), model_name='thenlper/gte-large', cache_folder=None, model_kwargs={}, encode_kwargs={}, multi_process=False)

In [ ]:
retriever1 = vectorstore1.as_retriever( #Converting the Chroma vector store into a retriever for querying.
    search_type='similarity', #Specifying that retrieval is based on cosine similarity
    search_kwargs={'k': 3} #Retrieving the top 3 most similar documents for a given query.
)

In [ ]:
def generate_rag_response1(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks1 = retriever1.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks1]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

temperature updated to 0.66 with chunk_overlap=20

In [ ]:
user_input1 = 'What is the protocol for managing sepsis in a critical care unit?'
print(generate_rag_response(user_input1,temperature=0.66))

Llama.generate: prefix-match hit


To manage sepsis in a critical care unit, healthcare professionals should follow these steps:
1. Suspect sepsis based on clinical signs and symptoms, such as fever, rapid heart rate, rapid breathing, low blood pressure, confusion or decreased level of consciousness.
2. Obtain cultures of blood, body fluids, and wound sites for Gram stain and culture before administering antibiotics.
3. Start prompt empiric antibiotic therapy immediately after suspecting sepsis. One regimen for septic shock of unknown cause is gentamicin or tobramycin 5.1 mg


top_k updated to 80 with chunk_overlap=0

In [ ]:
user_input1 = 'What is the protocol for managing sepsis in a critical care unit?'
print(generate_rag_response1(user_input1,top_k=80))

Llama.generate: prefix-match hit


Based on the context provided, the following are the steps for managing sepsis in a critical care unit:

1. Obtain specimens of blood, body fluids, and wound sites for Gram stain and culture before administering parenteral antibiotics.
2. Start very prompt empiric therapy immediately after suspecting sepsis.
3. Select an antibiotic regimen based on the suspected source, clinical setting, knowledge or suspicion of causative organisms and sensitivity patterns common to that specific inpatient unit, and previous culture results.
4. Consider adding vancomycin


#### Observations

* Response from generate_rag_response:
  * Style: Slightly more descriptive and explanatory in tone.

  * Focus: Emphasizes clinical suspicion and symptoms at the beginning.

  * Step 1: Starts with “Suspect sepsis based on clinical signs and symptoms…” – this makes it more patient-centered.

  * Step 3: Suggests a specific empiric antibiotic regimen for septic shock (gentamicin or tobramycin).

  * Includes more clinical detail, like vital signs (e.g., fever, rapid HR, hypotension).

  * Provides an example of empiric therapy, which is practical and actionable.

* Response from generate_rag_response1:
  * Style: Concise and directive.

  * Focus: Jumps directly to diagnostics (e.g., obtaining specimens).

  * Step 1: Starts with “Obtain specimens of blood…” – task-oriented, skipping clinical suspicion.

  * Additional Steps: Adds a fourth step not present in the first output: “Consider adding vancomycin.”

  * Lacks the introductory context (e.g., clinical signs).

  * More structured and protocol-driven with antibiotic selection criteria.

  * The content seems more extracted from reference-style material rather than generative inference.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

Fine tuning of retreiver parameters k=2

In [ ]:
retriever2 = vectorstore.as_retriever( #Converting the Chroma vector store into a retriever for querying.
    search_type='similarity', #Specifying that retrieval is based on cosine similarity
    search_kwargs={'k': 2} #Retrieving the top 2 most similar documents for a given query.
)

In [ ]:
def generate_rag_response2(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever2.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

In [ ]:
generate_rag_response2("What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?")

Llama.generate: prefix-match hit


'Appendicitis is characterized by abdominal pain, anorexia, and abdominal tenderness. The diagnosis is clinical, often supplemented by CT or ultrasound. Treatment involves surgical removal of the appendix. Antibiotics are used before surgery to prevent infection and after surgery if the appendix has perforated. If the appendix is not removed, complications such as gangrene, perforation, and peritonitis may occur. Therefore, medicine alone cannot cure appendicitis.'

In [ ]:
generate_rag_response("What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?",temperature=0.7)

Llama.generate: prefix-match hit


'The common symptoms for appendicitis include abdominal pain, anorexia (loss of appetite), and abdominal tenderness. Appendicitis cannot be cured through medication alone, and the standard treatment is surgical removal of the appendix, known as appendectomy.'

In [ ]:
generate_rag_response2("What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?",top_p=0.62)

Llama.generate: prefix-match hit


"Appendicitis is characterized by abdominal pain, anorexia (loss of appetite), and abdominal tenderness. It cannot be cured with medicine alone as the treatment involves surgical removal of the inflamed appendix. The surgery can be performed either through open or laparoscopic methods. Antibiotics are administered before the procedure to prevent infection. If the appendix is perforated, antibiotics should be continued until the patient's temperature and white blood cell count have normalized. In cases where a large inflammatory mass is found involving the appendix, terminal"

#### Observations

* Response from generate_rag_response2 (k=2)
  * Symptoms: abdominal pain, anorexia, abdominal tenderness

  * Extra Info: diagnosis via CT/ultrasound, antibiotics used pre-surgery, complications if untreated

  * Style: Balanced explanation with cause, diagnostic approach, and complications

  * Tone: Slightly more clinical, covers edge cases (gangrene, perforation, peritonitis)

  * Conclusion: Clear that medicine alone isn’t sufficient

* response from generate_rag_response(k=3, temperature=0.7)

  * Symptoms: abdominal pain, anorexia, abdominal tenderness

  * Conciseness: Most compact and direct of all 3

  * Treatment: States directly that medicine alone isn’t sufficient, surgery is the standard

  * Tone: Informative and clear, with minimum verbosity

* Response from generate_rag_response2 (k=2,top_p=0.62)

  * Symptoms: abdominal pain, anorexia, tenderness

  * More Detail:

  * Describes open vs laparoscopic surgery

  * Mentions continued antibiotic use in case of perforation

  * Ends mid-sentence, possibly due to truncation

  * Tone: Most exhaustive but verbose

  * Drawback: Likely pulled from a chunk that was too long, causing an incomplete output

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

Fine tuning LLM params Temperature to 0.7

In [ ]:
generate_rag_response("What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?",temperature=0.7)

Llama.generate: prefix-match hit


'The Merck Manual describes alopecia areata as a condition characterized by sudden patchy hair loss in people with no obvious skin or systemic disorder. It is most frequently seen on the scalp and beard but can affect any hairy area of the body. Alopecia areata is thought to be an autoimmune disorder affecting genetically susceptible individuals exposed to unclear environmental triggers.\n\nTreatments for alopecia areata include:\n1. Topical or intralesional corticosteroids: These drugs can stimulate hair growth and have been shown to be effective'

Fine tuning LLM params top_p=0.6

In [ ]:
generate_rag_response("What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?",top_p=0.6)

Llama.generate: prefix-match hit


'Based on the context provided, the effective treatments or solutions for addressing sudden patchy hair loss include:\n1. Corticosteroids (topical, intralesional, or systemic)\n2. Minoxidil (topical)\n3. Anthralin (topical)\n4. Immunotherapy (diphencyprone or squaric acid dibutylester)\n5. Psoralen plus ultraviolet A (PUVA)\n6. Hormonal modulators such as oral contraceptives or spironolactone for female-pattern hair'

#### Observations

* Response with temperature=0.7
  * Provides a clear explanation of the cause (autoimmune origin, genetic susceptibility, environmental triggers).

  * Describes alopecia areata as the condition behind the symptom.

  * Mentions diagnostic context (scalp and beard involvement, no skin/systemic disorder).

  * Lists only one treatment: topical or intralesional corticosteroids.

  * Tone: Balanced, narrative, clinically informative.

  * Limitation: Narrow treatment scope — lacks diversity in treatment options.

* Response with top_p=0.6
  * Skips cause explanation — jumps directly into treatment.

  * Lists six treatment options, covering:

  * Corticosteroids (topical/intralesional/systemic)

  * Minoxidil

  * Anthralin

  * Immunotherapy (DPCP, squaric acid dibutylester)

  * PUVA (phototherapy)

  * Hormonal therapy (e.g., spironolactone)

  * Tone: Directive, list-style, more like a clinical summary.

  * Strength: Comprehensive, practical for clinicians or patient education.

  * Limitation: No background explanation of the condition.

if we compare this with original response which was not finetuned that gives concise and superior response.  

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

Fine Tuning LLM params top_p=0.8

In [ ]:
generate_rag_response("What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?",top_p=0.8)

Llama.generate: prefix-match hit


'###Answer\nThe initial treatment includes ensuring a reliable airway and maintaining adequate ventilation, oxygenation, and blood pressure. Surgery may be needed in patients with more severe injuries to place monitors to track and treat intracranial pressure, decompress the brain if intracranial pressure is increased, or remove intracranial hematomas. In the first few days after the injury, maintaining adequate brain perfusion and oxygenation and preventing complications of altered sensorium are important. Subsequently, many patients require rehabilitation. There is no specific treatment for traumatic brain injury, but supportive care'

Fine Tuning LLM params temperature=0.1

In [ ]:
generate_rag_response("What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?",temperature=0.1)

Llama.generate: prefix-match hit


'###Answer\nThe initial treatment includes ensuring a reliable airway and maintaining adequate ventilation, oxygenation, and blood pressure. Surgery may be needed to place monitors to track and treat intracranial pressure, decompress the brain if intracranial pressure is increased, or remove intracranial hematomas. In the first few days after the injury, maintaining adequate brain perfusion and oxygenation and preventing complications of altered sensorium are important. Subsequently, many patients require rehabilitation. There is no specific treatment for brain damage, but supportive care should include preventing systemic complications due to'

#### Observations

* Response with top_p=0.8

  * Describes initial treatment steps: airway, ventilation, oxygenation, blood pressure.

  * Includes surgical interventions: monitoring, treating intracranial pressure, hematoma removal.

  * Highlights early-phase care: maintaining brain perfusion and preventing sensorium complications.

  * Mentions rehabilitation as a later-phase need.

  * Clearly states “no specific treatment” for traumatic brain injury — focus on supportive care.

  * Natural flow; reads like a clinical guideline.

  * Complete and informative without being overwhelming.

  * Balanced across emergency care, surgical needs, and rehabilitation.

  * Strong practical and clinical value.

  * Slightly less detailed at the end (trailing off after “supportive care”).

* Response with temperature=0.1

  * Very similar in structure and information as the first.

  * Includes all the same treatment domains (airway, pressure, perfusion, rehabilitation).

  * Adds more specifics in last line: “supportive care should include preventing systemic complications due to...” (but ends abruptly).

  * Deterministic and clinical.

  * Somewhat redundant or mechanical — expected due to low temperature.

  * Very controlled and consistent output.

  * Slightly more medically precise at the end.

  * Feels slightly repetitive due to lack of variation from low temperature.

original response without finetuning uses 95% probability which give more conside and succint answers.  

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

Fine Tuning LLM params with top_k=25

In [ ]:
generate_rag_response("What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?",top_k=25)

Llama.generate: prefix-match hit


'Based on the context provided, the following are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip:\n1. Elevate the injured limb above heart level for the first 2 days to minimize swelling and allow gravity to help drain edema fluid.\n2. Apply elastic bandages (batting, and layers 2 and 4) to the injured area after the initial 2 days.\n3. Periodically apply warmth using a heating pad for 15 to 20 minutes to relieve pain and speed healing.\n4.'

Fine Tuning LLM params with top_p=0.75

In [ ]:
generate_rag_response("What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?",top_p=0.75)

Llama.generate: prefix-match hit


'Based on the context provided, the following are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip:\n1. Elevate the injured limb above heart level for the first 2 days to minimize swelling and allow gravity to help drain edema fluid.\n2. Apply elastic bandages (batting, and layers 2 and 4) to the injured area after the initial 2 days.\n3. Periodically apply warmth using a heating pad for 15 to 20 minutes to relieve pain and speed healing.\n4.'

#### Observations

* Response from top_k=25

  * Parameter: Uses high recall (top_k=25), prioritizing breadth of retrieval from the context base.

  * Suggests elevation of the limb to reduce swelling.

  * Advises using elastic bandages after the initial 2 days.

  * Recommends applying warmth (heating pad) for pain relief and healing.

  * Ends with an incomplete line: \n4. — suggests there was more content intended but truncated or not retrieved/rendered.

  * Instructional, structured in a numbered list.

  * Despite high top_k, content richness didn’t increase significantly due to either truncation or limited relevant hits.

* Response from top_p=0.75

  * Parameter: Uses nucleus sampling (top_p=0.75), focusing on the most probable tokens from a cumulative distribution.

  * Identical in content and structure to the top_k=25 version — same 3 steps and same truncation at step 4.

  * Suffers the same cutoff issue with \n4. — incomplete answer.

  * Same clear, numbered format.

Original response without finetuning also generated the same response even when it has broader range of options top_k=50.   

## Output Evaluation

Rating the outputs(groudedness and relevance) so that evaluation can be comprehended easily.

In [ ]:
groundedness_rater_system_message = """

You will be presented a ###Question, ###Context used by the AI system and AI generated ###Answer.

Your task is to judge the extent to which the ###Answer is derived from ###Context.

Rate it 1 - if The ###Answer is not derived from the ###Context at all
Rate it 2 - if The ###Answer is derived from the ###Context only to a limited extent
Rate it 3 - if The ###Answer is derived from ###Context to a good extent
Rate it 4 - if The ###Answer is derived from ###Context mostly
Rate it 5 - if The ###Answer is is derived from ###Context completely

Note: Do not reward hallucinated or fabricated information. Only content directly traceable to the ###Context should be considered grounded.
Please note: Make sure you give a single overall rating in the range of 1 to 5 along with an overall explanation.

"""

In [ ]:
relevance_rater_system_message = """

You will be presented with a ###Question, the ###Context used by the AI system to generate a response, and the AI-generated ###Answer.

Your task is to judge the extent to which the ###Answer is relevant to the ###Question, considering whether it directly addresses the key aspects of the ###Question based on the provided ###Context.

Rate the relevance as follows:
- Rate 1 – The ###Answer is not relevant to the ###Question at all.
- Rate 2 – The ###Answer is only slightly relevant to the **###Question**, missing key aspects.
- Rate 3 – The ###Answer is moderately relevant, addressing some parts of the **###Question** but leaving out important details.
- Rate 4 – The ###Answer is mostly relevant, covering key aspects but with minor gaps.
- Rate 5 – The ###Answer is fully relevant, directly answering all important aspects of the **###Question** with appropriate details from the **###Context**.

Note: Do not reward hallucinated or fabricated information. Only content directly traceable to the ###Context should be considered grounded.

Note: Provide a single overall rating in the range of 1 to 5, along with a brief explanation of why you assigned that score.
"""

In [ ]:
user_message_template = """
###Question
{question}

###Context
{context}

###Answer
{answer}
"""

In [ ]:
def generate_ground_relevance_response(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=3)
    context_list = [d.page_content for d in relevant_document_chunks]
    context_for_query = ". ".join(context_list)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"""[INST]{qna_system_message}\n
                {'user'}: {qna_user_message_template.format(context=context_for_query, question=user_input)}
                [/INST]"""

    response = llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    answer =  response["choices"][0]["text"]

    # Combine user_prompt and system_message to create the prompt
    groundedness_prompt = f"""[INST]{groundedness_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    # Combine user_prompt and system_message to create the prompt
    relevance_prompt = f"""[INST]{relevance_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    response_1 = llm(
            prompt=groundedness_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    response_2 = llm(
            prompt=relevance_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    return response_1['choices'][0]['text'],response_2['choices'][0]['text']

In [ ]:
user_input1 = 'What is the protocol for managing sepsis in a critical care unit?'
groundedness,relevance = generate_ground_relevance_response(user_input1,max_tokens=128)

print(groundedness,end="\n\n")
print(relevance)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Rating: 5

Explanation: The answer is derived completely from the context. The context provides detailed information about the protocol for managing sepsis in a critical care unit, including giving parenteral antibiotics after taking specimens for Gram stain and culture, starting very prompt empiric therapy as soon as sepsis is suspected, and selecting antibiotics based on the suspected source, clinical setting, and previous culture results. The answer directly quotes the specific antibiotic regimen mentioned in the context for septic shock of unknown cause.

 Rating: 5

Explanation: The context provides detailed information about the management of critically ill patients in an ICU, including the importance of preventing infection and timely detection and treatment of sepsis. The answer directly addresses the question by specifying the protocol for managing sepsis in a critical care unit, which includes giving parenteral antibiotics after taking specimens for culture, starting prompt 

#### Observations

Both ratings being 5 implies the response is:

* Factual (grounded in source)

* On-topic (aligned with the user query)

* Complete (covers key components: detection, antibiotics, ICU steps)

* Excellent performance from the model in this context — both evaluation metrics validate a high-confidence answer.

In [ ]:
user_input2 = 'What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?'
groundedness,relevance = generate_ground_relevance_response(user_input2,max_tokens=128)

print(groundedness,end="\n\n")
print(relevance)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Rating: 5 - The Answer is derived from the Context completely.

Explanation: The Context clearly states that Appendicitis is a condition that requires surgical removal for treatment, and the common symptoms include abdominal pain, anorexia, and abdominal tenderness. The Answer accurately reflects this information by stating that the common symptoms are abdominal pain, anorexia, and abdominal tenderness, and that Appendicitis cannot be cured via medicine alone, with the standard treatment being surgical removal of the appendix through open or laparoscopic appendectomy.

 Rating: 5

Explanation: The answer directly addresses all important aspects of the question by listing the common symptoms for appendicitis and stating that it cannot be cured via medicine alone, but rather requires surgical removal. The answer also specifies the standard treatment as open or laparoscopic appendectomy. The context provided in the text fully supports these statements.


#### Observations

The response scores a perfect 5 on both groundedness and relevance.

* Complete Coverage: The model comprehensively handles a multi-part medical query without drifting off-topic.

* Factual Accuracy: The explanation shows strong grounding in source material — boosting trustworthiness.

* Implication: The LLM is performing optimally for complex medical queries, especially where context fidelity and relevance are critical.



In [ ]:
user_input3 = 'What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?'
groundedness,relevance = generate_ground_relevance_response(user_input3,max_tokens=128)

print(groundedness,end="\n\n")
print(relevance)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Rating: 5 - The Answer is derived from the Context completely. The context mentions various treatments for alopecia areata, including topical, intralesional, and systemic corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). The Answer repeats these treatments verbatim. Additionally, the context mentions hormonal modulators for female-pattern hair loss, which is also included in the

 Rating: 5

Explanation: The AI-generated answer directly addresses the question by providing a list of effective treatments and solutions for sudden patchy hair loss, which is a key aspect of the question. It also mentions alopecia areata specifically as a cause of sudden patchy hair loss, which aligns with the context provided. The answer also touches upon hormonal modulators for female-pattern hair loss, which is another potential cause mentioned in the context. Overall, the answer is fully relev

#### Observations

* Perfect scores again for both groundedness and relevance.

* Context alignment: Shows that the model understands medical terminology and maps symptoms to appropriate conditions (e.g., alopecia areata).

* Complete & focused: Effectively covers causes and treatments, ensuring completeness without over-explaining.

* Insight: The prompt includes a long and medically specific question, and the model handles it precisely with a grounded and relevant response.




In [ ]:
user_input4 = 'What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?'
groundedness,relevance = generate_ground_relevance_response(user_input4,max_tokens=350)

print(groundedness,end="\n\n")
print(relevance)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Rating: 5 - The AI generated answer is completely derived from the context. The context provides detailed information about the initial treatment for a person with a traumatic brain injury, including ensuring a reliable airway, maintaining adequate ventilation, oxygenation, and blood pressure, as well as the potential need for surgery to monitor intracranial pressure or remove hematomas. The context also mentions the importance of preventing complications in the first few days after the injury and the need for rehabilitation subsequently. The AI generated answer accurately reflects this information. Additionally, the context discusses the lack of a specific treatment for traumatic brain injury and the importance of supportive care, which is also included in the AI generated answer.

 Rating: 5

Explanation: The AI-generated answer directly addresses all important aspects of the question based on the provided context. It mentions the initial treatment for a person with a traumatic brai

#### Observations

* Faithful to context: No extraneous information was added; the model strictly adhered to the source.

* Precise and comprehensive: The model understood the medical complexity and responded with clarity and completeness.

* Semantic alignment: Effectively mapped “physical injury to brain tissue” → traumatic brain injury (TBI), showcasing medical reasoning.

In [ ]:
user_input5 = 'What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?'
groundedness,relevance = generate_ground_relevance_response(user_input5,max_tokens=350)

print(groundedness,end="\n\n")
print(relevance)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Rating: 5 - The Answer is derived from the Context completely.

Explanation: The context provides detailed information about the necessary precautions and treatment steps for a person with a fractured leg. The answer summarizes this information accurately and comprehensively, making it a direct reflection of the context.

 Rating: 5

Explanation: The answer directly addresses all the key aspects of the question regarding the necessary precautions and treatment steps for a person with a fractured leg. It covers elevating the injured limb, applying heat after 48 hours, immobilization using elastic bandages or a cast, inspecting the cast, avoiding putting objects inside the cast, seeking medical care if needed, and the importance of good hygiene. The answer also mentions potential complications such as severe swelling leading to compartment syndrome and prolonged immobilization causing stiffness, contractures, and muscle atrophy. Overall, the answer is fully relevant to the question base

#### Observations

* Comprehensive: Covers both immediate first-aid and post-care.

* Faithful to context: Directly maps to the context's structure and terminology.

* Semantically accurate: Links a fractured leg during a hike to standard field care protocols.



## Actionable Insights and Business Recommendations


* **Business Insights from RAG-Based Medical Assistant**
  * Accuracy & Trustworthiness
    * RAG-based (Retrieval-Augmented Generation) systems deliver highly reliable, context-specific answers, significantly boosting user confidence and satisfaction—especially critical in medical applications where accuracy is paramount.

  * Enhanced User Experience
    * By generating responses that are both personalized and highly relevant, the RAG model improves user engagement and satisfaction, offering a seamless and intelligent interaction experience.

  * Cost Optimization in Healthcare
    * These models can automate medical Q&A processes, reducing the dependency on live consultations. This leads to substantial cost savings in telemedicine, customer support, and triage services.

* **Recommendations**
  * Continuous Fine-Tuning
    * To maintain relevance and accuracy, the model should be regularly updated with the latest medical research, clinical guidelines, and real-world data.

  * Ethical and Regulatory Compliance
    * Ensure strict adherence to data privacy standards such as HIPAA and GDPR, especially when handling sensitive patient information.

  * Specialized Domain Models
    * Develop dedicated models tailored to individual medical fields (e.g., orthopedics, gastroenterology, oncology) to further enhance domain-specific accuracy and contextual depth.

<font size=6 color='blue'>Power Ahead</font>
___

In [ ]:
!pip install numpy==1.25.2 -q

In [ ]:
import os

os.kill(os.getpid(), 9)